allennlp/modules/similarity_functions/linear.py

```diff
@@ -0,0 +1,143 @@
+from typing import Callable
+
+from overrides import overrides
+import torch
+from torch.nn.parameter import Parameter
+
+from allennlp.common import Params
+from allennlp.common.checks import ConfigurationError
+from allennlp.experiments import Registry
+from allennlp.modules import SimilarityFunction
+
+
+@Registry.register_similarity_function("linear")
+class LinearSimilarity(SimilarityFunction):
+    """
+    This similarity function performs a dot product between a vector of weights and some
+    combination of the two input vectors, followed by an (optional) activation function.  The
+    combination done is configurable.
+
+    If the two vectors are ``x`` and ``y``, we allow the following kinds of combinations: ``x``,
+    ``y``, ``x*y``, ``x+y``, ``x-y``, ``x/y``, where each of those binary operations is performed
+    elementwise.  You can list as many combinations as you want, comma separated.  For example, you
+    might give ``x,y,x*y`` as the ``combination`` parameter to this class.  The computed similarity
+    function would then be ``w^T [x; y; x*y] + b``, where ``w`` is a vector of weights, ``b`` is a
+    bias parameter, and ``[;]`` is vector concatenation.
+
+    Note that if you want a bilinear similarity function with a diagonal weight matrix W, where the
+    similarity function is computed as `x * w * y + b` (with `w` the diagonal of `W`), you can
+    accomplish that with this class by using "x*y" for `combination`.
+
+    Parameters
+    ----------
+    tensor_1_dim : ``int``
+        The dimension of the first tensor, ``x``, described above.  This is ``x.size()[-1]`` - the
+        length of the vector that will go into the similarity computation.  We need this so we can
+        build weight vectors correctly.
+    tensor_2_dim : ``int``
+        The dimension of the second tensor, ``y``, described above.  This is ``y.size()[-1]`` - the
+        length of the vector that will go into the similarity computation.  We need this so we can
+        build weight vectors correctly.
+    combination : ``str``, optional (default=``"x,y"``)
+        Described above.
+    activation : ``Callable[[torch.Tensor], torch.Tensor]``, optional (default=``lambda x: x``)
+        An activation function applied after the ``w^T * [x;y] + b`` calculation.  Default is no
+        activation.
+    """
+    def __init__(self,
+                 tensor_1_dim: int,
+                 tensor_2_dim: int,
+                 combination: str = 'x,y',
+                 activation: Callable[[torch.Tensor], torch.Tensor] = lambda x: x) -> None:
```

Comment:
This output is very likely going directly into a softmax, and I wouldn't want to pass those through a `tanh` or most other activation functions first, because then you make it pretty much impossible for the softmax to be peaky.  I think the 'linear' default activation function (same as the default activation for Keras' `Dense` layer) is reasonable here.